src/coreComponents/constitutive/fluid/multifluid/CO2Brine/functions/CO2Solubility.hpp

```diff
@@ -229,112 +241,169 @@ CO2SolubilityUpdate::compute( real64 const & pressure,
+    watSolubilityDeriv[ic] *= m_componentMolarWeight[m_CO2Index];
   }
 
-  // Y = C/W = z/(1-z)
-  real64 Y = 0.0;
-  real64 dY_dCompFrac[2]{};
+  real64 const z_co2 = compFraction[m_CO2Index];
+  real64 const z_wat = compFraction[m_waterIndex];
 
-  if( compFraction[m_CO2Index] > 1.0 - minForDivision )
+  real64 const determinant = 1.0 - co2Solubility*watSolubility;
+  real64 invDeterminant = 0.0;
+  real64 invDeterminantDeriv[] = { 0.0, 0.0 };
+  if( minForDivision < LvArray::math::abs( determinant ) )
   {
-    Y = compFraction[m_CO2Index] / minForDivision;
-    dY_dCompFrac[m_CO2Index] = 1.0 / minForDivision;
-    dY_dCompFrac[m_waterIndex] = 0.0;
+    invDeterminant = 1.0 / determinant;
+    for( integer ic = 0; ic < 2; ic++ )
+    {
+      invDeterminantDeriv[ic] = invDeterminant*invDeterminant*(co2Solubility*watSolubilityDeriv[ic] + watSolubility*co2SolubilityDeriv[ic]);
+    }
   }
   else
   {
-    real64 const oneMinusCompFracInv = 1.0 / (1.0 - compFraction[m_CO2Index]);
-    Y = compFraction[m_CO2Index] * oneMinusCompFracInv;
-    dY_dCompFrac[m_CO2Index] = oneMinusCompFracInv * oneMinusCompFracInv;
-    dY_dCompFrac[m_waterIndex] = 0.0;
+    invDeterminant = 1.0 / minForDivision;
   }
 
-  auto setZero = []( real64 & val ){ val = 0.0; };
-  LvArray::forValuesInSlice( phaseFraction.derivs, setZero );
-  LvArray::forValuesInSlice( phaseCompFraction.derivs, setZero );
-
-  if( Y < solubility )
+  real64 x_co2 = co2Solubility * (z_wat - watSolubility * z_co2) * invDeterminant;
+  real64 x_co2Deriv[4]{ 0.0, 0.0, 0.0, 0.0 };
+  if( minForDivision < x_co2 )
   {
-    // liquid phase only
-
-    // 1) Compute phase fractions
-
-    phaseFraction.value[m_phaseLiquidIndex] = 1.0;
-    phaseFraction.value[m_phaseGasIndex] = 0.0;
-
-    // 2) Compute phase component fractions
-
-    phaseCompFraction.value[m_phaseGasIndex][m_CO2Index] = 1.0;
-    phaseCompFraction.value[m_phaseGasIndex][m_waterIndex] = 0.0;
-    for( localIndex ic = 0; ic < 2; ++ic )
+    // Pressure and temperature derivatives
+    for( integer ic = 0; ic < 2; ic++ )
     {
-      phaseCompFraction.value[m_phaseLiquidIndex][ic] = compFraction[ic];
+      x_co2Deriv[ic] = co2SolubilityDeriv[ic] * (z_wat - watSolubility * z_co2) * invDeterminant
+                       - co2Solubility * watSolubilityDeriv[ic] * z_co2 * invDeterminant
+                       + co2Solubility * (z_wat - watSolubility * z_co2) * invDeterminantDeriv[ic];
+    }
+    // Composition derivatives
+    x_co2Deriv[2+m_CO2Index] = -co2Solubility * watSolubility * invDeterminant;
+    x_co2Deriv[2+m_waterIndex] = co2Solubility * invDeterminant;
+  }
+  else
+  {
+    x_co2 = 0.0;
+  }
 
-      for( localIndex jc = 0; jc < 2; ++jc )
-      {
-        phaseCompFraction.derivs[m_phaseLiquidIndex][ic][Deriv::dC+jc] = (ic == jc ) ? 1.0 : 0.0;
-        phaseCompFraction.derivs[m_phaseGasIndex][ic][Deriv::dC+jc] = 0.0;
-      }
+  real64 y_wat = watSolubility * (z_co2 - x_co2);
```

Comment:
Yes this is a substitution. This is because the value of x_co2 can potentially be truncated to ensure a minimum value of 0 - negative values can occur if z_wat is very small. Such cases would reduce to single phase anyway.

As for which is best to start solving for, my take is that it will be a fairly stable 2x2 linear system with `determinant ~= 1` so it really shouldn't matter in most cases.